### ISSUE TITLE
Faces.sendFile() inside p:columns logs jakarta.servlet.ServletException

### ISSUE THREAD

**Issue by BryanV91** [omnifaces/omnifaces#801]
Issue:

I'm trying to download a file inside a `p:dataTable`, `p:fileDownload` works normally but `Faces.sendFile()` downloads the file and throws a warning in the console:

```
[WARNING] /test.xhtml
jakarta.servlet.ServletException: cannot set buffer size after response is in committed state
	at jakarta.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:709)
	at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:449)
	at org.eclipse.jetty.ee10.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1370)
	at org.eclipse.jetty.ee10.servlet.ServletHolder.handle(ServletHolder.java:736)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1614)
	at org.eclipse.jetty.ee10.websocket.servlet.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:195)
	at org.eclipse.jetty.ee10.servlet.FilterHolder.doFilter(FilterHolder.java:205)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1586)
	at org.eclipse.jetty.ee10.servlet.ServletHandler$MappedServlet.handle(ServletHandler.java:1547)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.dispatch(ServletChannel.java:819)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:431)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:464)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:575)
	at org.eclipse.jetty.ee10.servlet.SessionHandler.handle(SessionHandler.java:703)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:851)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:151)
	at org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:619)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:411)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalStateException: cannot set buffer size after response is in committed state
	at org.eclipse.jetty.ee10.servlet.ServletApiResponse.setBufferSize(ServletApiResponse.java:387)
	at com.sun.faces.context.ExternalContextImpl.setResponseBufferSize(ExternalContextImpl.java:876)
	at jakarta.faces.context.ExternalContextWrapper.setResponseBufferSize(ExternalContextWrapper.java:932)
	at jakarta.faces.context.ExternalContextWrapper.setResponseBufferSize(ExternalContextWrapper.java:932)
	at org.omnifaces.util.FacesLocal.sendFile(FacesLocal.java:1919)
	at org.omnifaces.util.FacesLocal.sendFile(FacesLocal.java:1948)
	at org.omnifaces.util.FacesLocal.sendFile(FacesLocal.java:1874)
	at org.omnifaces.util.Faces.sendFile(Faces.java:2807)
	at org.primefaces.test.TestView.download(TestView.java:54)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.el.parser.AstValue.invoke(AstValue.java:253)
	at org.apache.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:267)
	at org.jboss.weld.module.web.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
	at org.jboss.weld.module.web.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
	at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:70)
	at jakarta.faces.event.MethodExpressionActionListener.processAction(MethodExpressionActionListener.java:107)
	at jakarta.faces.event.ActionEvent.processListener(ActionEvent.java:72)
	at jakarta.faces.component.UIComponentBase.broadcast(UIComponentBase.java:480)
	at jakarta.faces.component.UICommand.broadcast(UICommand.java:197)
	at jakarta.faces.component.UIData.broadcast(UIData.java:1035)
	at jakarta.faces.component.UIData.broadcast(UIData.java:1035)
	at jakarta.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:858)
	at jakarta.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:999)
	at com.sun.faces.lifecycle.ApplyRequestValuesPhase.execute(ApplyRequestValuesPhase.java:51)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:72)
	at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:131)
	at jakarta.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:691)
	... 29 more

```

```
<h:form id="frmTest">
<p:dataTable id="tbl" var="record" value="#{testView.list}">
	<p:columns id="column" value="#{testView.cols}" var="col" columnIndexVar="index" rendered="#{col.rendered}">
		<p:commandButton value="Download" ajax="false" icon="pi pi-arrow-down" styleClass="mr-2">
			<p:fileDownload value="#{testView.file}" /> File #{column}
		</p:commandButton>
		<h:commandLink id="file-btn" immediate="true" target="_blank"
			actionListener="#{testView.download()}">
			omnifaces - download
		</h:commandLink>
	</p:columns>
</p:dataTable>
</h:form>
```

Not sure if this issue belongs here, but this used to work prior to updating all my environment:

- Java 8 to 17
- Mojarra 2.3 to 4.0
- PrimeFaces 13.0.1 to 13.0.8 Jakarta
- OmniFaces 3.1 to 4.4

Attaching reproducer: 

[primefaces-test-jakarta.zip](https://github.com/omnifaces/omnifaces/files/15017349/primefaces-test-jakarta.zip)

Make sure to change the path in this function to whatever file you have in your local computer.

```
public void download() throws IOException {
    Faces.sendFile(new File("change me"), false);
}
```

**Comment by pizzi80** [comment]
reproduced also with Tomcat 10.1.20

I think that is a Mojarra bug, because the method `download()` it's called 3 times!

The second and third time when the `FacesLocal.sendFile` method 
call `externalContext.setResponseBufferSize(DEFAULT_SENDFILE_BUFFER_SIZE)`

the underlying `HttpServletResponse` throws an `IllegalStateException`

**Comment by pizzi80** [comment]
it's the combination of PrimeFaces datatable
if you try whith an h:datatable the action is called only once.

To me it's still a Mojarra bug

```
<h:dataTable id="tbl2" var="record" value="#{testView.list}">
    <c:forEach items="#{testView.cols}" var="col">
        <h:column rendered="#{col.rendered}">
            <p:commandButton id="pfDownload2" value="Download" ajax="false" icon="fa fa-download" styleClass="mr-2">
                <p:fileDownload value="#{testView.file}" /> File #{column}
            </p:commandButton>
            <!-- target="_blank" -->
            <h:commandButton id="downloadJsf" action="#{testView.download}" value="Jsf download" />
            <h:commandLink id="downloadJsfLink" action="#{testView.download}" value="Jsf download js" />
            <h:commandLink id="foo" action="#{testView.foo}" value="Jsf Foo" />
        </h:column>
    </c:forEach>
</h:dataTable>
```

Try also with a `foo()` method

```
class TestView {
 
  public void foo() {
     log.info("foo()");
  }
}

```

**Comment by pizzi80** [comment]
removing `p:columns` it's working also with `p:dataTable`

```
<p:dataTable id="tbl3" var="record" value="#{testView.list}">
  <c:forEach items="#{testView.cols}" var="col">
      <p:column rendered="#{col.rendered}">
          <p:commandButton id="pfDownload" value="Download" ajax="false" icon="fa fa-download" styleClass="mr-2">
              <p:fileDownload value="#{testView.file}" /> File #{column}
          </p:commandButton>
          <!-- target="_blank" -->
          <h:commandButton id="jsfDownloadButton" action="#{testView.download}" value="Jsf download" />
          <h:commandLink id="jsfDownloadLink" action="#{testView.download}" value="Jsf download js" styleClass="button" />
          <h:commandLink id="jsfFoo" action="#{testView.foo}" value="Jsf Foo" styleClass="button" />
      </p:column>
  </c:forEach>
</p:dataTable>


```

**Comment by BryanV91** [comment]
Thanks for the response, I can confirm that `Faces.sendFile()` works well outside of `p:columns`. I will report this issue to PrimeFaces team.

**Comment by melloware** [comment]
@BryanV91 please do and make sure to include an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.

**Comment by melloware** [comment]
This can be closed it is fixed in PF 13.0.9 and 14.0.0
